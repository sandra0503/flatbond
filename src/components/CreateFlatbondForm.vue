<template>
  <div class="hello">
    <h1>Welcome to flatfair!</h1>
    <p>Enter your rent and post code to create your personal Flatbond.</p>
  </div>
</template>

<script>
export default {
  name: "CreateFlatbondForm"
};
</script>

<style scoped lang="scss"></style>
