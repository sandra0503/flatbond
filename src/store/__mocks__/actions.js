export default {
  FETCH_CONFIG: jest.fn(),
  CREATE_FLATBOND: jest.fn().mockResolvedValue("created"),
  SET_FEE: jest.fn()
};
