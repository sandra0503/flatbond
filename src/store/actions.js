import { fetchConfig, createFlatbond } from "../api.js";

async function FETCH_CONFIG({ commit }) {
  try {
    const config = await fetchConfig();
    commit("SET_CONFIG", config);
  } catch (error) {
    console.log(error);
  }
}

function CREATE_FLATBOND({ commit }) {
  return new Promise(async (resolve, reject) => {
    try {
      const status = await createFlatbond();
      console.log("actions.js: config", status);
      commit("SET_FLATBOND", status);
      resolve(status);
    } catch (error) {
      reject(error);
    }
  });
}

export default {
  FETCH_CONFIG,
  CREATE_FLATBOND
};
