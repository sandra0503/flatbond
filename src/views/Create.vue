<template>
  <div class="create">
    <div class="create__introduction">
      <h1>Welcome to flatfair!</h1>
      <p>
        To create your personal flatbond, we need some details about your new
        home.
      </p>
    </div>
    <CreateFlatbondForm
      @submitted="createFlatbond"
      v-if="isMembershipFeeConfigPresent"
    />
    <Illustration :imageSrc="require('@/assets/house.png')" />
  </div>
</template>

<script>
import { mapState } from "vuex";
import CreateFlatbondForm from "@/components/CreateFlatbondForm.vue";
import Illustration from "@/components/Illustration.vue";

export default {
  name: "create",
  components: {
    CreateFlatbondForm,
    Illustration
  },
  computed: {
    ...mapState(["feeConfig"]),
    isMembershipFeeConfigPresent() {
      return (
        this.feeConfig && typeof this.feeConfig.fixedMembershipFee === "boolean"
      );
    }
  },
  created() {
    this.$store.dispatch("FETCH_CONFIG");
  },
  methods: {
    createFlatbond(flatbondData) {
      const { rent, postcode, fee } = flatbondData;

      this.$store
        .dispatch("CREATE_FLATBOND", {
          rent: rent,
          postcode: postcode
        })
        .then(() => {
          this.$store.dispatch("SET_FEE", fee);
          this.$router.push("/success");
        })
        .catch(() => {
          this.error = true;
        });
    }
  }
};
</script>

<style lang="scss" scoped>
.create {
  text-align: center;
  max-width: 1024px;
  padding-top: 30px;
  margin: 0 auto;
  padding-bottom: 200px;
}

.create__introduction {
  margin-bottom: 25px;
  max-width: 480px;
  margin: auto;
}
</style>
