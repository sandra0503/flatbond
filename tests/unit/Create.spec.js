jest.mock("@/store/actions");
import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import Create from "@/views/Create";
import CreateFlatbondForm from "@/components/CreateFlatbondForm";
import initialState from "@/store/state";
import actions from "@/store/actions";

const localVue = createLocalVue();
localVue.use(Vuex);

describe("Create view", () => {
  let state;

  const build = () => {
    const wrapper = shallowMount(Create, {
      localVue,
      store: new Vuex.Store({
        state,
        actions
      })
    });

    return {
      wrapper,
      createFlatbondForm: () => wrapper.find(CreateFlatbondForm)
    };
  };

  beforeEach(() => {
    state = { ...initialState };
  });

  it("renders the component and matches snapshot", () => {
    const { wrapper } = build();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("does not render the form component when config is not present", () => {
    const { createFlatbondForm } = build();

    expect(createFlatbondForm().exists()).toBe(false);
  });

  it("renders the form component", () => {
    state.feeConfig = {
      fixedMembershipFee: false,
      fixedMembershipFeeAmount: "12000"
    };

    const { createFlatbondForm } = build();

    expect(createFlatbondForm().exists()).toBe(true);
  });

  it("fetches the configuration", () => {
    expect(actions.FETCH_CONFIG).toHaveBeenCalled();
  });

  it("creates flatbond when received 'submitted'", () => {
    state.feeConfig = {
      fixedMembershipFee: false,
      fixedMembershipFeeAmount: "12000"
    };

    const { createFlatbondForm } = build();

    const flatbondData = {
      rent: 20000,
      postcode: "WC2N 5DU",
      fee: 137
    };

    createFlatbondForm().vm.$emit("submitted", flatbondData);

    expect(actions.CREATE_FLATBOND).toHaveBeenCalled();
    expect(actions.CREATE_FLATBOND.mock.calls[0][1]).toEqual({
      rent: 20000,
      postcode: "WC2N 5DU"
    });
  });
});
