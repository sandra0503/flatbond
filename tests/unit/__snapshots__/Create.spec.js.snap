// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create view renders the component and matches snapshot 1`] = `
<div class="create">
  <div class="create__introduction">
    <h1>Welcome to flatfair!</h1>
    <p>
      To create your personal flatbond, we need some details about your new
      home.
    </p>
  </div>
  <!---->
  <illustration-stub imagesrc="../assets/house.png"></illustration-stub>
</div>
`;
